examples/boot_counter/flash_sector.rs

```diff
@@ -0,0 +1,89 @@
+use core::{
+    mem::{size_of, MaybeUninit},
+    ptr::read_volatile,
+};
+
+use rp2040_flash::flash;
+
+/// XIP base address (see `XIP_BASE` in RP2040 datasheet).
+pub const FLASH_ORIGIN: usize = 0x10000000;
+/// The erasable sector size.
+const FLASH_SECTOR_SIZE: usize = 4096;
+/// The value an erased sector is filled with. This is typically 0xff.
+const FLASH_ERASED_VALUE: u8 = 0xff;
+
+/// The payload type `T` must fit into a single flash sector.
+///
+/// The payload type should be `repr(C)` to have a stable layout,
+/// because the flash-stored payload can survive firmware upgrades.
+pub union FlashSector<T>
+where
+    T: Copy,
+{
+    data: [u8; FLASH_SECTOR_SIZE],
+    value: MaybeUninit<T>,
+}
+
+impl<T> Default for FlashSector<T>
+where
+    T: Copy,
+{
+    fn default() -> Self {
+        Self {
+            data: [FLASH_ERASED_VALUE; FLASH_SECTOR_SIZE],
+        }
+    }
+}
+
+impl<T> FlashSector<T>
+where
+    T: Copy,
+{
+    pub fn new(value: T) -> Self {
+        assert!(
+            size_of::<T>() <= FLASH_SECTOR_SIZE,
+            "`T` must fit into a single sector size"
+        );
+        let mut instance = Self::default();
+        instance.value = MaybeUninit::new(value);
+        instance
+    }
+
+    pub fn read(mem_addr: usize) -> Self {
+        assert!(
+            size_of::<T>() <= FLASH_SECTOR_SIZE,
+            "`T` must fit into a single sector size"
+        );
+        assert!(mem_addr >= FLASH_ORIGIN);
+
+        let mut flash_sector = FlashSector::default();
+        flash_sector.value = unsafe { read_volatile(mem_addr as *const _) };
+        flash_sector
+    }
+
+    pub unsafe fn write(&self, mem_addr: usize) {
+        assert!(mem_addr >= FLASH_ORIGIN);
+
+        let flash_addr = mem_addr - FLASH_ORIGIN;
+        assert!(is_aligned(flash_addr, FLASH_SECTOR_SIZE));
+
+        flash::flash_range_erase_and_program(flash_addr as u32, &self.data, true);
+    }
+
+    pub fn data(&self) -> &[u8] {
+        unsafe { &self.data }
+    }
+
+    pub fn value(&self) -> MaybeUninit<T> {
+        unsafe { self.value }
+    }
+}
+
+const fn is_aligned(addr: usize, alignment: usize) -> bool {
+    assert!(is_pwr_of_two(alignment));
+    addr & (alignment - 1) == 0
+}
+
+const fn is_pwr_of_two(n: usize) -> bool {
```

Comment:
There already is https://doc.rust-lang.org/std/primitive.usize.html#method.is_power_of_two which you could use.